File: Regular-Mocks /DSA/RemoveZeroSumNodes.py
# PS: https://leetcode.com/problems/remove-zero-sum-consecutive-nodes-from-linked-list/?envType=daily-question&envId=2024-03-12

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    # 
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        front = ListNode(0, head)
        start = front 
        prefixSumMap = {}
        prefixSum = 0
        # Calculate teh prefix sum and store them into map
        while start:
            prefixSum += start.val
            prefixSumMap[prefixSum] = start
            start = start.next

        start = front
        prefixSum = 0

        while start:
            prefixSum += start.val
            start.next = prefixSumMap[prefixSum].next
            start = start.next
        
        return front.next

    
    
    
    # Do a simple bruteforce
    # T.C: O(N^2), S.C: O(1)
    def removeZeroSumSublistsBruteForce(self, head: Optional[ListNode]) -> Optional[ListNode]:
        front = ListNode(0, head) # dummy-node
        start = front

        while start:
            prefixSum = 0
            end = start.next
            
            # Fix the start and then move end to the end of the list.
            while end:
                prefixSum += end.val
                if prefixSum == 0:
                    start.next = end.next
                end = end.next
            start = start.next
        
        return front.next


        
        

File: Regular-Mocks /DSA/RemoveZeroSumNodes.java
// PS: https://leetcode.com/problems/remove-zero-sum-consecutive-nodes-from-linked-list

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 /*
    [1, 2, 3, -3, -2]
    [1  3  6   3   1]

    [1, 2, -3]
    [1, 3, 0]

    [1, 2, 3, -3, -2]
    [1, 3, 6, 3, 1]

    T.C: O(N), S.C: O(N)

 */
class Solution {

    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode front = new ListNode(0, head); // Check this if it fails
        Map<Integer, ListNode> map = new HashMap();
        int prefixSum = 0;
        ListNode curr = front;
        
        // Fill in the prefix map
        while(curr != null){
            prefixSum += curr.val;
            map.put(prefixSum, curr);
            curr = curr.next;
        }
        
        curr = front;
        prefixSum = 0;

        while(curr != null){
            prefixSum += curr.val;
            // map[prefixSum].next ;
            curr.next = map.get(prefixSum).next;
            curr = curr.next;
        }

        return front.next;
    }
}